### ISSUE TITLE
Incorrect gradient with atan2 and "late" masking

### ISSUE THREAD

**Issue by jonahthelion** [pytorch/pytorch#93998]
### 🐛 Describe the bug

Masking invalid values before applying `atan2` gives the correct gradient, but masking afterwards puts a `nan` in the gradient. The correct gradient in this case is `tensor([-0.1333,  0.0000,  0.0000,  0.0000])`, not `tensor([-0.1333,     nan, -0.0000, -0.0000])`.

I think in general people mask out padding only just before evaluating the loss, which is the order that gives the incorrect `nan` in this case, so this could be an important issue.

```
import torch

x = torch.Tensor([1,0,2,3])
y = torch.Tensor([2,0,0,0])
k = torch.BoolTensor([True, False, True, True])
x.requires_grad = True
out = torch.atan2(y[k], x[k])
out.mean().backward()
print(x.grad)
# -> tensor([-0.1333,  0.0000,  0.0000,  0.0000])

x = torch.Tensor([1,0,2,3])
y = torch.Tensor([2,0,0,0])
k = torch.BoolTensor([True, False, True, True])
x.requires_grad = True
out = torch.atan2(y, x)
out[k].mean().backward()
print(x.grad)
# -> tensor([-0.1333,     nan, -0.0000, -0.0000])
```

### Versions

Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.5.2 (x86_64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.7 | packaged by conda-forge | (default, Sep 14 2021, 01:18:03)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-11.5.2-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.1
[pip3] numpy==1.21.2
[pip3] torch==1.13.1
[pip3] torchvision==0.2.2
[conda] blas                      2.111                       mkl    conda-forge
[conda] blas-devel                3.9.0              11_osx64_mkl    conda-forge
[conda] efficientnet-pytorch      0.7.1                    pypi_0    pypi
[conda] libblas                   3.9.0              11_osx64_mkl    conda-forge
[conda] libcblas                  3.9.0              11_osx64_mkl    conda-forge
[conda] liblapack                 3.9.0              11_osx64_mkl    conda-forge
[conda] liblapacke                3.9.0              11_osx64_mkl    conda-forge
[conda] mkl                       2021.3.0           h08c4f10_555    conda-forge
[conda] mkl-devel                 2021.3.0           h694c41f_556    conda-forge
[conda] mkl-include               2021.3.0           h08c4f10_555    conda-forge
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torchvision               0.2.2                      py_3    pytorch

**Comment by kta-intel** [comment]
Thanks for pointing this out. I am able to reproduce this issue using cuda on Ubuntu 20.04.5 LTS. We will look into this

**Comment by albanD** [comment]
Hi,

I'm afraid this is the expected behavior. The short version is that, at points where the function is not differentiable, AD doesn't work anymore. And so things like commutativity don't hold.
There are a lot of discussions about this in https://github.com/pytorch/pytorch/issues/4132 if you want more details.
